### ISSUE TITLE
Cancel is not working (API error: Invalid or missing task id"

### ISSUE THREAD

**Issue by DanielHillebrandSWMS** [duplicati/duplicati#4453]
- [ ] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.5.1_beta_2020-01-18
- **Operating system**: Windows 10 Pro
- **Backend**: Local folder (USB drive)

## Description

I have tried to cancel a restore process, but no reaction after clicking on cancel button. 

I am clicking here:
![grafik](https://user-images.githubusercontent.com/48449212/108875553-9a401a00-75fd-11eb-9140-c6f251b1d491.png)

With dev-tools I can see that the API endpoint returns an error message: 

![grafik](https://user-images.githubusercontent.com/48449212/108874610-9a8be580-75fc-11eb-9391-fc3f5de31b85.png)

## Steps to reproduce

1. Create new backup config with destination on disk
2. Execute backup
3. Restore backup to a new destination (temp folder)
4. Wait until in state "Scannen nach lokalen Blöcken" (scan for local blocks?)
4. Click on "Cancel" (German "abbrechen" in my case)

- **Actual result**:
no response
- **Expected result**:
cancel restore

## Screenshots

Funny enough the task id seems correct (an example of the progress-state answers:

![grafik](https://user-images.githubusercontent.com/48449212/108875041-11c17980-75fd-11eb-9f46-4f517f454efc.png)

Addition: It just tried again (different task id!) and have seen that by clicking the X in the progress bar and confirming the security question stopping works:

![grafik](https://user-images.githubusercontent.com/48449212/108875849-e4290000-75fd-11eb-866a-0d0338dee8a3.png)

**Comment by warwickmm** [comment]
I wasn't able to reproduce this error with version 2.0.5.112.  However, in my case it didn't actually cancel the restore (perhaps I was too slow for the small backup I tested with).

There were a few fixes for stopping backups since 2.0.5.1, but I'm not sure if those affected cancelling a restore process.

**Comment by ts678** [comment]
> it didn't actually cancel the restore

This is consistent with the "no reaction after clicking on cancel button".

> There were a few fixes for stopping backups

Some seem to have involved changing the `IsCancellationRequested` testing -- which seem totally absent in RestoreHandler.cs.
I'm not sure if another method was used, or if cancel was never coded. Though title mentions error, the wish was "cancel restore".

**Comment by warwickmm** [comment]
> This is consistent with the "no reaction after clicking on cancel button".

Thanks @ts678.  I forgot to mention that I did not notice any 404 errors in any POST requests.  Unlike the screenshot that shows "stop", i see "stopnow" and "stopaftercurrentfile", depending on which button I clicked.  I don't remember what options were available in 2.0.5.1.

**Comment by DanielHillebrandSWMS** [comment]
Hi @warwickmm , thanks for the new title. Far better than what I came up with  ;-)

The "stopnow" worked. 

"stopnow" is called when clicking on the x in the circle
"stop" is called when clicking the "cancel" button.

**Comment by warwickmm** [comment]
Thanks @DanielHillebrandSWMS.  That is my fault, I didn't read your description carefully and was clicking the wrong button.  I can reproduce this now.

![image](https://user-images.githubusercontent.com/17345532/109427064-7dad3480-79a5-11eb-937d-cc4527d7c8b7.png)

I don't know what the purpose is of the cancel button.  If it's unnecessary, we can remove it.

**Comment by DanielHillebrandSWMS** [comment]
You are very welcome @warwickmm. I am glad that you could reproduce it.

**Comment by ts678** [comment]
I think the cancel of a restore might leave a huge mess of partially restored files, some with full length but hidden gaps inside, because the restore is done a dblock at a time, and every dblock might hold blocks from many files. Restore distributes those.

Because Duplicati checks an existing target file for blocks that don't need updating, it's possible a second restore will clean up, however somebody should test that. I'm not sure if that thinking factored into current design, but it might guide future work.

What's the motivation for cancel? I'm worried that people will complain that their files (maybe the original paths) got trashed.

**Comment by warwickmm** [comment]
I agree, it's not clear what should happen when cancelling a restore. Perhaps that's why it's not really implemented?  I think the file hash is verified after a file is restored, so perhaps there's some hope that a few files are restored fully.

**Comment by Xavron** [comment]
Confused by what you're saying and what I see with Duplicati.

Clicking the x brings up the stop now action and that seems to be basically what one can view as a cancel. The files remain and the restore is stopped or aka cancelled.

The cancel button doesn't seem to do anything or at least doesn't cancel it in a way people expect so you could probably just make it do what stop now does. To me, that would be far less confusing.

**Comment by warwickmm** [comment]
Thanks @Xavron.  You're right, we do handle task cancellation for restores (older implementation, but it still works).  My earlier testing gave the appearance that the cancellation was being ignored because it was able to find all the blocks from local source files and I couldn't cancel fast enough.

I think it's best to remove the cancel button, as having another mechanism to interrupt the task is just confusing.

**Comment by DanielHillebrandSWMS** [comment]
Just to explain why I hit cancel in the first place: I am a first-time Duplicati user and was testing the restore. The backup I was restoring was very big and in the first try I only wanted to see that it worked it all, but did't want to wait for all the files to be restored.

@Xavron Files should not be deleted on cancellation. Maybe only the current file if the process can be stopped mid-file so you won't have a file that is not complete. No sure.

@warwickmm Just removing the cancel button sounds good to me. 

Thank you!
